src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java

```diff
@@ -299,14 +374,113 @@ protected int doWork() {
+     * @param sam an open {@link SamReader} from which to pull reads.
+     * @param referenceSequenceFileWalker an open {@link ReferenceSequenceFileWalker} from which to get reference sequence information.
+     * @return An open {@link SamLocusAndReferenceIterator} ready to iterate.
+     */
+    private SamLocusAndReferenceIterator createSamLocusAndReferenceIterator(final SamReader sam, final ReferenceSequenceFileWalker referenceSequenceFileWalker) {
+
+        sequenceDictionary = referenceSequenceFileWalker.getSequenceDictionary();
+        if (sam.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+            throw new PicardException("Input BAM must be sorted by coordinate");
+        }
+
+        // Make sure our reference and reads have teh same sequence dictionary:
+        sequenceDictionary.assertSameDictionary(sam.getFileHeader().getSequenceDictionary());
+
+        final IntervalList regionOfInterest = getIntervals(sequenceDictionary);
+
+        log.info("Getting SamLocusIterator");
+
+        final SamLocusIterator samLocusIterator = new SamLocusIterator(sam, regionOfInterest);
+
+        // Now go through and compare information at each of these sites
+        samLocusIterator.setEmitUncoveredLoci(false);
+        samLocusIterator.setMappingQualityScoreCutoff(MIN_MAPPING_Q);
+        samLocusIterator.setQualityScoreCutoff(MIN_BASE_Q);
+
+        log.info("Using " + aggregatorList.size() + " aggregators.");
+
+        aggregatorList.forEach(la ->
+                IOUtil.assertFileIsWritable(new File(OUTPUT + la.getSuffix())));
+
+        // iterate over loci
+        log.info("Starting iteration over loci");
+
+        final SamLocusAndReferenceIterator iterator = new SamLocusAndReferenceIterator(referenceSequenceFileWalker, samLocusIterator);
+
+        // This hasNext() call has side-effects. It loads up the index and makes sure that
+        // the iterator is really ready for
+        // action. Calling this allows for the logging to be more accurate.
+        iterator.hasNext();
+        return iterator;
+    }
+
+    /**
+     * Initializes the counts, random object, and iterators for this class.
+     *
+     * Specifically, the following fields are initialized:
+     *
+     *   random
+     *   aggregatorList
+     *   progressLogger
+     *   nTotalLoci
+     *   nSkippedLoci
+     *   nProcessedLoci
+     */
+    private void initializeAggregationState() {
+        random = new Random(42);
+
+        aggregatorList = getAggregatorList();
+
+        progressLogger = new ProgressLogger(log, PROGRESS_STEP_INTERVAL);
+        nTotalLoci = 0;
+        nSkippedLoci = 0;
+        nProcessedLoci = 0;
+    }
+
+
+    @Override
+    protected int doWork() {
+
+        // Initialize our iteration:
+        initializeAggregationState();
+
+        // Make sure we can read our files:
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFilesAreReadable(INTERVALS);
```

Comment:
might as well assert that VCF is readable.

Reply:
and reference too

Reply:
Fixed.